sympy/integrals/heurisch.py

```diff
@@ -732,7 +732,7 @@ def find_non_syms(expr):
             return candidate.xreplace(solution).xreplace(
                 dict(zip(poly_coeffs, [S.Zero]*len(poly_coeffs))))
 
-    if not (F.free_symbols - set(V)):
+    if F.has_free(V, all=True):
```

Comment:
I don't think this is equivalent. `F.free_symbols - set(V)` is the set of symbols in `F` that are *not* in `V`. Then `not (F.free_symbols - set(V))` means "F has no symbols that are not in V". However `F.has_free(V, all=True)` means that "F has all the symbols that are in V".

This change reverses the sense of the subsets so before it was
```python
if F.free_symbols <= V
```
but the changed code is basically
```python
if V <= F.free_symbols
```

Reply:
So this is a case where has can't help. We need to know what are all the free generators are are in F...and we need to know what qualifies as a generator. Another way to look at this would be `F.xreplace(dict(zip (V, (Dummy() for I in V)))).free_symbols & F.free_symbols` is empty.

Having spent many hours on the two approaches, I can say that it is very disconcerting to have to think about the types of generators that one must consider at the algorithm level. I would be MUCH in favor of replacing generators with symbols at the routine outset.

Reply:
I wonder what the purpose of the check is in the first place. Maybe @asmeurer would know.